File: 2017_05_09_CLV/dataLoad/BigQuery_DataLoad.R
# LOADING DATA FROM BIGQUERY

library("bigrquery")
library("data.table")


# List available projects, datasets and tables

list_projects() # This asks you to authorize bigrquery in the browser

project = "clv-research-paveljasek"

list_datasets(project)

list_tables(project, dataset = "clv_aa")


# DATA LOAD

#sql = "select * from [clv-research-paveljasek:clv_aa.clv_transactions_input]"
#data_df = query_exec(sql, project = project, max_pages = Inf)

data_df = list_tabledata("clv-research-paveljasek", "clv_aa", "clv_transactions_input", max_pages = Inf)

head(data_df, n = 10)
hist(data_df$profit, xlim = c(0, 3000), breaks = 500)


# DATA TRANSFORMATIONS

data = data.table(data_df)

prevMonday <- function(x)
  7 * floor(as.numeric(x - 1 + 4) / 7) + as.Date(1 - 4, origin = "1970-01-01")

data[, monday_date := prevMonday(as.Date(date))]
data = data[order(date, transaction)]


# a) daily

data_daily = data[, .(transaction = transaction[1]
                      , date_user_created = date_user_created[1]
                      , date_first_order = date_first_order[1]
                      , zip_firstchar = zip_firstchar[1]
                      , profit = sum(profit)
                      , channel_poe = channel_poe[1]
                      , channel_type = channel_type[1]
                      , medium_source = medium_source[1]
                      , transaction_revenue = sum(transaction_revenue)
                      , item_quantity = sum(item_quantity)
                      , transaction_shipping = sum(transaction_shipping)
                      , count = .N)
                  , by = .(customer, date)]


# b) weekly

data_weekly = data[, .(transaction = transaction[1]
                       , date_user_created = date_user_created[1]
                       , date_first_order = date_first_order[1]
                       , zip_firstchar = zip_firstchar[1]
                       , profit = sum(profit)
                       , channel_poe = channel_poe[1]
                       , channel_type = channel_type[1]
                       , medium_source = medium_source[1]
                       , transaction_revenue = sum(transaction_revenue)
                       , item_quantity = sum(item_quantity)
                       , transaction_shipping = sum(transaction_shipping)
                       , count = .N)
                   , by = .(customer, monday_date)]


# DATA CHECK

sum(data$profit)
sum(data_daily$profit)
sum(data_weekly$profit)

File: 2017_05_09_CLV/dataLoad/BigQuery_DataLoad.py
# LOADING DATA FROM BIGQUERY

from google.cloud import bigquery # We use this library only to list available datasets...
import pandas as pd # ... the real querying is done directly with pandas!
import datetime


# List available datasets and tables

client = bigquery.Client(project = 'clv-research-paveljasek')

for dataset in client.list_datasets():
    print(dataset.path)
    
dataset = client.dataset('clv_aa')

for table in dataset.list_tables():
    print(table.path)


# DATA LOAD

sql = "select * from [clv-research-paveljasek:clv_aa.clv_transactions_input]"

data = pd.read_gbq(query = sql, project_id = 'clv-research-paveljasek')


# DATA TRANSFORMATIONS

def prevMonday(x):
    x = datetime.datetime.strptime(x, "%Y-%m-%d").date()
    mon = x - datetime.timedelta(days = x.weekday())
    return(mon.strftime("%Y-%m-%d"))
    
data['monday_date'] = data['date'].apply(lambda x: prevMonday(x))
data.sort_values(by = ['date', 'transaction'], inplace = True)

# a) daily

aggregations = {
    'transaction': {'transaction': 'first'
                    , 'count': 'count'}
    , 'date_user_created': {'date_user_created': 'first'}
    , 'date_first_order': {'date_first_order': 'first'}
    , 'zip_firstchar': {'zip_firstchar': 'first'}
    , 'profit': {'profit': 'sum'}
    , 'channel_poe': {'channel_poe': 'first'}
    , 'channel_type': {'channel_type': 'first'}
    , 'medium_source': {'medium_source': 'first'}
    , 'transaction_revenue': {'transaction_revenue': 'sum'}
    , 'item_quantity': {'item_quantity': 'sum'}
    , 'transaction_shipping': {'transaction_shipping': 'sum'}
    }

data_daily = data.groupby(['customer', 'date']).agg(aggregations)
data_daily.columns = data_daily.columns.get_level_values(1)
data_daily['customer'] = data_daily.index.get_level_values(0)
data_daily['date'] = data_daily.index.get_level_values(1)
data_daily.reset_index(drop = True, inplace = True)


# b) weekly

data_weekly = data.groupby(['customer', 'monday_date']).agg(aggregations)
data_weekly.columns = data_weekly.columns.get_level_values(1)
data_weekly['customer'] = data_weekly.index.get_level_values(0)
data_weekly['monday_date'] = data_weekly.index.get_level_values(1)
data_weekly.reset_index(drop = True, inplace = True)


# DATA CHECK

data['profit'].sum()
data_daily['profit'].sum()
data_weekly['profit'].sum()
